state-chain/runtime/src/chainflip/bitcoin_elections.rs

```diff
@@ -299,56 +290,279 @@ impl Hook<btc::BlockNumber, Vec<DepositChannelDetails<Runtime, BitcoinInstance>>
 	}
 }
 
-pub struct BitcoinDepositChannelWitessingProcessor;
+type Age = BlockNumber;
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, TypeInfo, Deserialize, Serialize)]
+pub enum BtcEvent {
+	PreWitness(BlockNumber, DepositWitness<Bitcoin>),
+	Witness(BlockNumber, DepositWitness<Bitcoin>),
+}
+impl BtcEvent {
+	pub fn deposit_witness(&self) -> &DepositWitness<Bitcoin> {
+		match self {
+			BtcEvent::PreWitness(_, dw) | BtcEvent::Witness(_, dw) => dw,
+		}
+	}
+	pub fn equal_inner(&self, other: &BtcEvent) -> bool {
+		self.deposit_witness() == other.deposit_witness()
+	}
+}
 
-impl ProcessBlockData<btc::BlockNumber, Vec<DepositWitness<Bitcoin>>>
-	for BitcoinDepositChannelWitessingProcessor
+/// Returns one event per deposit witness. If multiple events share the same deposit witness:
+/// - keep the `Witness` variant,
+/// - otherwise keep the single `PreWitness`.
+pub fn deduplicate_btc_events(events: Vec<BtcEvent>) -> Vec<BtcEvent> {
+	// Map: deposit_witness -> chosen BtcEvent
+	let mut chosen: BTreeMap<DepositWitness<Bitcoin>, BtcEvent> = BTreeMap::new();
+
+	for event in events {
+		let deposit: DepositWitness<Bitcoin> = event.deposit_witness().clone();
+
+		match chosen.get(&deposit) {
+			None => {
+				// No event yet for this deposit, store it
+				chosen.insert(deposit, event);
+			},
+			Some(existing) => {
+				// There's already an event for this deposit
+				match (existing, &event) {
+					// If we already have a Witness, do nothing
+					(BtcEvent::Witness(_, _), BtcEvent::PreWitness(_, _)) => (),
+					// If we have a PreWitness and the new event is a Witness, override it
+					(BtcEvent::PreWitness(_, _), BtcEvent::Witness(_, _)) => {
+						chosen.insert(deposit, event);
+					},
+					// This should be impossible to reach!
+					(_, _) => (),
+				}
+			},
+		}
+	}
+
+	chosen.into_values().collect()
+}
+
+#[derive(
+	Debug,
+	Clone,
+	PartialEq,
+	Eq,
+	Encode,
+	Decode,
+	TypeInfo,
+	Deserialize,
+	Serialize,
+	MaxEncodedLen,
+	Default,
+)]
+enum BtcRuleType {
+	PreWitness,
+	Witness,
+	#[default]
+	None,
+}
+
+#[derive(
+	Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+pub struct DepositChannelWitessingProcessor<ChainBlockNumber: Ord, BlockData, Event> {
+	pub blocks_data: BTreeMap<ChainBlockNumber, (BlockData, ChainBlockNumber)>,
+	pub reorg_events: BTreeMap<ChainBlockNumber, Vec<Event>>,
+	pub rules: BoundedVec<BtcRuleType, ConstU32<2>>,
```

Comment:
The rules shouldn't change during the runtime, so we should ensure that they don't have to be stored in storage.
We can encode the rules by adding them as another associated type in the `BWProcessorTypes`:
```
...
   type Rules : Hook<(Self::ChainBlockNumber, Age), Self::Event>;
```
We'll still have a `rules` field here, but it will only contain data when running tests and will be empty when building for real.

```suggestion
        pub rules: Rules,
```